File: mq-demos/amqp_server.rb
# ruby amqp_server.rb
# based on https://www.rabbitmq.com/tutorials/tutorial-six-ruby.html

require 'bunny'
require 'json'

class RpcServer
  def initialize
    @connection = Bunny.new
    @connection.start
    @channel = @connection.create_channel
  end

  def start(queue_name)
    @queue = channel.queue(queue_name)
    @exchange = channel.default_exchange
    subscribe_to_queue
  end

  def stop
    channel.close
    connection.close
  end

  def loop_forever
    # This loop only exists to keep the main thread
    # alive. Many real world apps won't need this.
    loop { sleep 5 }
  end

  private

  attr_reader :channel, :exchange, :queue, :connection

  def subscribe_to_queue
    queue.subscribe do |_delivery_info, properties, payload|
      request = JSON.parse(payload.encode('utf-8'))
      puts request
      result = get_length request

      exchange.publish(
        JSON.generate(result).encode('utf-8'),
        routing_key: properties.reply_to,
        correlation_id: properties.correlation_id
      )
    end
  end

  def get_length(request)
    length = request['string'].length
    {:length => length}
  end
end

begin
  server = RpcServer.new

  puts 'Awaiting RPC requests'
  server.start('rpc_queue')
  server.loop_forever
rescue Interrupt => _
  server.stop
end

File: mq-demos/amqp_server.py
# python3 amqp_server.py
# based on https://www.rabbitmq.com/tutorials/tutorial-six-python.html
import pika
import json


connection = pika.BlockingConnection(
    pika.ConnectionParameters(host='localhost'))
channel = connection.channel()
channel.queue_declare(queue='rpc_queue')


def get_length(request):
    length = len(request['string'])
    return {'length': length}


def on_request(ch, method, props, body):
    try:
        request = json.loads(body.decode('utf-8'))
    except json.decoder.JSONDecodeError:
        ch.basic_ack(delivery_tag=method.delivery_tag)
        print('Bad request:', body)
        return
    
    print(request)    
    response = get_length(request)
    body = json.dumps(response).encode('utf-8')
    
    ch.basic_publish(exchange='',
                     routing_key=props.reply_to,
                     properties=pika.BasicProperties(correlation_id=props.correlation_id),
                     body=body)
    ch.basic_ack(delivery_tag=method.delivery_tag)

channel.basic_qos(prefetch_count=1)
channel.basic_consume(queue='rpc_queue', on_message_callback=on_request)

print("Awaiting RPC requests")
channel.start_consuming()